File: ALGO/Python/mergeSort.py
def merge(a, left, mid, right):
    i = left
    j = mid+1
    aux = []
    while i <= mid and j <= right:
        if a[i] < a[j]:
            aux.append(a[i])
            i += 1
        else:
            aux.append(a[j])
            j += 1
    while i <= mid:
        aux.append(a[i])
        i += 1
    while j <= right:
        aux.append(a[j])
        j += 1
    a = a[:left] + aux + a[right+1:]
    # print(aux)
    return a


def merge_sort(a, left, right):
    if left == right:
        return a
    mid = (left+right)//2
    a = merge_sort(a, left, mid)
    a = merge_sort(a, mid+1, right)
    return merge(a, left, mid, right)


if __name__ == '__main__':
    arr = [12, 34, 56, 23, 78, 38, 43, 89]
    arr = merge_sort(arr, 0, len(arr)-1)
    print(arr)

File: ALGO/C/Sort/mergeSort.c
#include <stdio.h>
#include <stdlib.h>

static int aux[10];

int intCmp(int *x, int *y)
{
    printf("%d %d\n", *x, *y);
    return *(x) < *(y);
}

void merge(int *arr, int l, int mid, int r)
{
    int i = l;
    int j = mid + 1;
    int k = 0;
    while (i <= mid && j <= r)
        if (intCmp(arr+i,arr+j))
            aux[k++] = arr[i++];
        else
            aux[k++] = arr[j++];
    while (j <= r)
        aux[k++] = arr[j++];
    while (i <= mid)
        aux[k++] = arr[i++];
    for (i = 0; i < k; i++)
        arr[l + i] = aux[i];
}

void mergeSort(int *arr, int l, int r)
{
    if (l == r)
        return;
    int mid = (l + r) / 2;
    mergeSort(arr, l, mid);
    mergeSort(arr, mid + 1, r);
    merge(arr, l, mid, r);
}

int main()
{
    // int arr[] = {56, 23, 45, 54, 76, 98, 34, 54, 12, 1};
    // mergeSort(arr, 0, 9);
    int arr[] = {5,2,1,4,3};
    mergeSort(arr,0,4);
    int i;
    for (i = 0; i < 10; i++)
        printf("%d ", arr[i]);
    return 0;
}
